### ISSUE TITLE
explorer.exe is consuming lots of CPU while client is running and idle

### ISSUE THREAD

**Issue by spasche** [haiwen/seafile-client#886]
I'm using Seafile Client 6.0.4 on Windows 10.
I noticed that the explorer.exe process is continuously consuming lots of CPU while Seafile is running. Even if there is no activity inside the Seafile folder.
In the following screenshot, we can see that explorer.exe is consuming about 30% CPU. Once I stop the Seafile client, CPU usage drops to around 0%:
![2017-03-28](https://cloud.githubusercontent.com/assets/105751/24418871/b972103e-13ed-11e7-9423-a9a03d5c8aae.png)

I tried disabling Explorer Extension in the settings, but it has no effect.
I have around 270000 files inside my libraries.

**Comment by killing** [comment]
Are the files inside the library frequently changed by some program? How much CPU does seaf-daemon.exe uses?

**Comment by spasche** [comment]
No, nothing I'm aware of is updating files. seaf-daemon.exe CPU usage stays low at around 2.5%.

**Comment by mhoyer** [comment]
I partially can confirm this noticeable CPU load. Not that high on my site, but still 4-5%, when `seaf-daemon.exe` is running (of course, after indexing). `seaf-daemon.exe` itself is also constantly requiring 1-3% CPU usage. It is actually not much, but on my notebook I can hear my CPU fan blowing all the time (esp. now in summer :-)).

![image](https://user-images.githubusercontent.com/76611/27011540-e971b68a-4ebd-11e7-870a-5cf776f79079.png)

![image](https://user-images.githubusercontent.com/76611/27011559-1d5dc4d4-4ebe-11e7-810f-3cafdd0381f8.png)

`tail -n 3 ~/ccnet/logs/*.log` also shows no recent entries.

My setup:
- Lenovo X220, i7, 16GB Ram, SSD
- Win10 x64
- Seafile Client 6.0.6
- A single library: ~7GB, ~550k files in 97k folders

Is there any chance of enabling additional logging?

**Comment by gschintgen** [comment]
Hi, I can confirm this issue. It's getting out of hand and is killing my laptop's battery life:

With seafile client running (but without activity in any of the libraries) Windows Explorer has a mostly constant CPU usage of around 15% (whereas seaf-daemon itself only consumes some 2%-3% of CPU time).
Disabling "auto sync" via right-click does not have any noticeable effect! (I find this quite strange.)
As soon as I properly quit Seafile Client the CPU *total* comes down to around 3%! So the culprit is definitely Seafile or its interaction with some Windows subsystem(s).

I can't see how this issue could be considered as a minor one. Some 15%+3% for an idle machine is horrendous.

Here's some more information about my system:
- latest available client (6.1.8)
- 3 synchronized libraries (10.7GB in 15000 files / 4GB in 4300 files / 7.3GB in 133000 files)
- latest Win10 (1809) x64
- Thinkpad with i7-4510U and 8GiB of RAM, SSD for storage

There's no constant activity logged to the logfiles.

I'd kindly suggest you raise the importance of this issue.

**Comment by gschintgen** [comment]
Does the client by any chance use a FileSystemWatcher? (That seems to be some kind of Windows equivalent to Linux's inotify). I found multiple reports that this can lead to similar problems:

[MSDN FileSystemWatcher high cpu load](https://social.msdn.microsoft.com/Forums/office/en-US/d219394c-caf5-4767-a2b2-db7c22339165/filesystemwatcher-high-cpu-load-when-no-changes-happen-to-the-monitored-directory-sub-directories?forum=netfxbcl)

[100% CPU usage when using FileSystemWatcher](https://social.msdn.microsoft.com/Forums/vstudio/en-US/25a7f812-e176-4fb5-a6d0-0c2f4966a791/100-cpu-usage-when-using-filesystemwatcher?forum=csharpgeneral)

[Why does my filesystemwatcher use so much cpu useage?](https://www.codeproject.com/Questions/1213500/Why-does-my-filesystemwatcher-use-so-much-cpu-usea)

Apparently a FileSystemWatcher uses polling (instead of listening to events emitted by the filesystem as it happens) and periodically/frequently rescans the watched folders in their entirety! Obviously this won't scale at all.

Would it be possible to add a delay somewhere in order to slow down this rescanning madness? (As also alluded to in some of the discussions linked above.)

But then again I don't know the first thing about developing on and for Windows...

**Comment by gschintgen** [comment]
> With seafile client running (but without activity in any of the libraries) Windows Explorer has a mostly constant CPU usage of around 15% (whereas seaf-daemon itself only consumes some 2%-3% of CPU time).

I'd like to add that, after all, it's *not* a constant load on the CPU. After quite some time it's finally settling down to minor background activity. So it may be related to the size (and number of items) in the libraries. Still, it's quite annoying.

**Comment by brintal** [comment]
I have the same issue. High CPU (20-30%) load when seafile client is running on my windows machine. As it is an older computer, the problem is quite intense as its not just making my cpu work a bit more, it actually takes up to 5 seconds to open a new explorer window or open a folder. 
I can also confirm that the problem seems to become less noticeable the longer the pc runs.

**Comment by Grendel13G** [comment]
I have exactly this issue with the same symptoms:

-Windows 10, client 7.0.0
-High CPU usage (primarily Windows Explorer, secondarily seaf-daemon.exe/Seafile) after all libraries have been synced and no changes are being made to library files
-Disabling auto sync, or unsyncing individual libraries, appears to have no effect
-Quitting Seafile immediately drops CPU usage to normal
-CPU usage eventually returns to normal after Seafile has been running for a long time (not sure how long, but the transition is sudden)

I even completely reinstalled Windows yesterday and have only synced my smaller libraries to try to alleviate this problem, but it persists.

**Comment by Rotzbua** [comment]
Same on Windows 8.1,  client 7.0.0 and previous.

Process Monitor shows that seafile does every second a `QueryDirectory` operation. This operation is even done if sync is paused.

Is there a option to increase the `QueryDirectory` operation aka polling?

**Comment by killing** [comment]
@Rotzbua The QueryDirectory system call should be called by seaf-daemon to check whether the library folder still exists. This shouldn't consume a lot of CPU.

We'll try to locate the problem.

**Comment by Rotzbua** [comment]
@killing Thanks. Some additional information:

> starting seaf-daemon:  ("-c", "C:/Users/XXX/ccnet", "-d", "C:/Users/XXX/Desktop/Seafile/seafile-data", "-w", "C:/Users/XXX/Desktop/Seafile")

Main seafile folder is on Desktop. Also on the Desktop some other single synced folders.

Screenshot
![abc](https://user-images.githubusercontent.com/7337347/59432190-06a5c600-8de7-11e9-9b93-7379ad2e299e.png)

**Comment by killing** [comment]
At least one cause of this issue is found to be seaf-daemon.exe will scan the entire library on restart and will refresh the files' status icon in Windows explorer. This will cause explorer to consume a lot of CPU when there are a lot of files in the library. This will be fixed in upcoming 7.0.6 version.
